language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,346 @@
+    /// everytime they create a new ballot. This is used for creating unique
+    /// global identifiers for Ballots
+    struct BallotCounter has key {
+        counter: u64,
+    }
+
+    /// RemoveBallotEvent is emitted when a ballot is
+    /// created by a proposer
+    struct CreateBallotEvent<Proposal: store + drop> has drop, store {
+        ballot_id: BallotID,
+        ballot: Ballot<Proposal>,
+    }
+
+    /// RemoveBallotEvent is emitted when a ballot has
+    /// been removed. This can either happen because:
+    /// * ballot was approved
+    /// * ballot was manually removed by the proposer
+    /// * ballot was expired and garbage collected
+    struct RemoveBallotEvent has drop, store {
+        ballot_id: BallotID,
+    }
+
+    /// VotedEvent is emitted when a valid vote has
+    /// been accepted by the ballot
+    struct VotedEvent has drop, store {
+        ballot_id: BallotID,
+        voter: address,
+        vote_weight: u64,
+    }
+
+    /// BallotApprovedEvent is emitted when a ballot has
+    /// been approved by the voters
+    struct BallotApprovedEvent has drop, store {
+        ballot_id: BallotID,
+    }
+
+    /// The maximum number of ballots allowed per proposal type
+    /// per address.
+    const MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS: u64 = 256;
+
+    /// The provided timestamp(s) were invalid
+    const EINVALID_TIMESTAMP: u64 = 1;
+    /// The address already contains has the maximum of ballots allowed
+    /// MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS
+    const ETOO_MANY_BALLOTS: u64 = 2;
+    /// Ballot with the provided id was not found
+    const EBALLOT_NOT_FOUND: u64 = 3;
+    /// Proposal details in the vote do not match the proposal details
+    /// in the ballot
+    const EBALLOT_PROPOSAL_MISMATCH: u64 = 4;
+    /// Voter not allowed to vote in the ballot
+    const EINVALID_VOTER: u64 = 5;
+    /// current timestamp < ballot start time
+    const EBALLOT_NOT_STARTED: u64 = 6;
+    /// current timestamp > ballot expiration time
+    const EBALLOT_EXPIRED: u64 = 7;
+    /// Voter has already voted in this ballot
+    const EALREADY_VOTED: u64 = 8;
+
+    /// A constructor for BallotID
+    public fun new_ballot_id(
+        counter: u64,
+        proposer: address,
+    ): BallotID {
+        BallotID {
+            counter,
+            proposer,
+        }
+    }
+
+    /// Create a ballot under the signer's address and return the `BallotID`
+    public fun create_ballot<Proposal: store + copy + drop>(
+        ballot_account: &signer,
+        proposal: Proposal,
+        proposal_type: vector<u8>,
+        num_votes_required: u64,
+        allowed_voters: vector<vector<u8>>,
+        start_timestamp_secs: u64,
+        expiration_timestamp_secs: u64
+    ): BallotID acquires Ballots, BallotCounter {
+        let ballot_address = Signer::address_of(ballot_account);
+        // Remove any expired ballots
+        // TODO gc_internal<Proposal>(ballot_address);
```

Comment:
Why is it TODO? We do have `gc_internal()` ready, don't we?
Btw, I'd propose to combine GC with the first time initialization of the ballots resource into some `setup()` function.

Reply:
I am getting a VM error here. Need to talk to @tzakian  on how to fix this.

`gc_internal()`  is calling `borrow_global_mut<Ballots<Proposal>>(addr)` and `create_ballot` is also calling `borrow_global_mut<Ballots<Proposal>>(addr)` with the same addr - looks like this is not allowed in move.

From the looks of it, `gc_internal()` does not release the mut pointer it borrowed.

> Btw, I'd propose to combine GC with the first time initialization of the ballots resource into some setup() function.

But setup is different, gc is something we want to call on every invocation of `create_ballot` and `setup` is something which we would only invoke once.

Reply:
Fixed the gc_internal issue.